src/main/java/com/tabnine/statusBar/TabnineStatusBarWidget.java

```diff
@@ -1,31 +1,53 @@
 package com.tabnine.statusBar;
 
+import com.intellij.openapi.application.ApplicationManager;
+import com.intellij.openapi.application.ModalityState;
 import com.intellij.openapi.editor.colors.EditorColorsManager;
 import com.intellij.openapi.project.Project;
 import com.intellij.openapi.wm.CustomStatusBarWidget;
+import com.intellij.openapi.wm.StatusBar;
+import com.intellij.openapi.wm.WindowManager;
 import com.intellij.openapi.wm.impl.status.EditorBasedWidget;
 import com.intellij.openapi.wm.impl.status.TextPanel;
 import com.intellij.util.Consumer;
 import com.tabnine.binary.BinaryRequestFacade;
 import com.tabnine.binary.requests.config.ConfigRequest;
+import com.tabnine.binary.requests.config.StateResponse;
 import com.tabnine.binary.requests.statusBar.ConfigOpenedFromStatusBarRequest;
+import com.tabnine.general.StaticConfig;
+import com.tabnine.lifecycle.BinaryStateChangeNotifier;
+import com.tabnine.lifecycle.BinaryStateService;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 
 import javax.swing.*;
+import java.awt.*;
 import java.awt.event.MouseAdapter;
 import java.awt.event.MouseEvent;
 import java.util.Objects;
 
 import static com.tabnine.general.StaticConfig.ICON_AND_NAME;
 import static com.tabnine.general.StaticConfig.ICON_AND_NAME_DARK;
 
-public class TabnineStatusBarWidget extends EditorBasedWidget implements CustomStatusBarWidget, com.intellij.openapi.wm.StatusBarWidget.WidgetPresentation {
+public class TabnineStatusBarWidget extends EditorBasedWidget implements CustomStatusBarWidget,
+        com.intellij.openapi.wm.StatusBarWidget.WidgetPresentation {
     private final BinaryRequestFacade binaryRequestFacade;
+    private volatile StateResponse.ServiceLevel serviceLevel;
+    private TextPanel.WithIconAndArrows component;
 
     public TabnineStatusBarWidget(@NotNull Project project, BinaryRequestFacade binaryRequestFacade) {
         super(project);
         this.binaryRequestFacade = binaryRequestFacade;
+        final BinaryStateService binaryStateService = ApplicationManager.getApplication().getComponent(BinaryStateService.class);
+        if (binaryStateService != null) {
+            StateResponse lastStateResponse = binaryStateService.getLastStateResponse();
+            if (lastStateResponse != null) {
+                this.serviceLevel = lastStateResponse.getServiceLevel();
+            }
+
+        }
```

Comment:
This is so that we can immediately display the widget with the latest known service level. Since the service only notifies on changes, if the service level stays the same, the widget will never get notified.

Reply:
I must admit I also did not understand the complexity here (why do we need both persist & notify on state), should it only solve the first init call?

Reply:
O.k, due to popular demand I changed this to take the value from the service when needed.